rholang/src/main/scala/coop/rchain/rholang/interpreter/Interpreter.scala

```diff
@@ -25,48 +26,55 @@ object Interpreter {
 
-  implicit def interpreter[F[_]](implicit F: Sync[F]): Interpreter[F] = new Interpreter[F] {
+  implicit def interpreter[F[_]](
+      implicit S: Sync[F],
+      C: _cost[F],
+      E: _error[F]
+  ): Interpreter[F] =
+    new Interpreter[F] {
 
-    def evaluate(runtime: Runtime[F], term: String): F[EvaluateResult] =
-      evaluate(runtime, term, Cost.Max)
+      def evaluate(runtime: Runtime[F], term: String): F[EvaluateResult] =
+        evaluate(runtime, term, Cost.Max)
 
-    def evaluate(runtime: Runtime[F], term: String, initialPhlo: Cost): F[EvaluateResult] = {
-      implicit val rand: Blake2b512Random = Blake2b512Random(128)
-      for {
-        checkpoint <- runtime.space.createCheckpoint()
-        res        <- injAttempt(runtime.reducer, runtime.errorLog, term, initialPhlo)
-        _          <- if (res.errors.nonEmpty) runtime.space.reset(checkpoint.root) else F.unit
-      } yield res
-    }
+      def evaluate(runtime: Runtime[F], term: String, initialPhlo: Cost): F[EvaluateResult] = {
+        implicit val rand: Blake2b512Random = Blake2b512Random(128)
+        for {
+          checkpoint <- runtime.space.createCheckpoint()
+          res        <- injAttempt(runtime.reducer, runtime.errorLog, term, initialPhlo)
+          _          <- if (res.errors.nonEmpty) runtime.space.reset(checkpoint.root) else S.pure(())
+        } yield res
+      }
 
-    def injAttempt(
-        reducer: ChargingReducer[F],
-        errorLog: ErrorLog[F],
-        term: String,
-        initialPhlo: Cost
-    )(implicit rand: Blake2b512Random): F[EvaluateResult] = {
-      val parsingCost = accounting.parsingCost(term)
-      // TODO: charge(parsingCost) so that it is visible in the cost log
-      val phloAfterParsing = initialPhlo - parsingCost
-      if (phloAfterParsing.value <= 0)
-        EvaluateResult(parsingCost, Vector(OutOfPhlogistonsError)).pure[F]
-      else
-        ParBuilder[F].buildNormalizedTerm(term).attempt.flatMap {
-          case Right(parsed) =>
-            for {
-              _         <- reducer.setPhlo(phloAfterParsing)
-              result    <- reducer.inj(parsed).attempt
-              phlosLeft <- reducer.phlo
-              oldErrors <- errorLog.readAndClearErrorVector()
-              newErrors = result.swap.toSeq.toVector
-              allErrors = oldErrors |+| newErrors
-            } yield EvaluateResult(initialPhlo - phlosLeft, allErrors)
-          case Left(error) =>
-            for {
-              phlosLeft <- reducer.phlo
-            } yield EvaluateResult(parsingCost, Vector(error))
-        }
-    }
+      def injAttempt(
+          reducer: ChargingReducer[F],
+          errorLog: ErrorLog[F],
+          term: String,
+          initialPhlo: Cost
+      )(implicit rand: Blake2b512Random): F[EvaluateResult] = {
+        val parsingCost = accounting.parsingCost(term)
+        for {
+          _           <- reducer.setPhlo(initialPhlo)
+          parseResult <- charge[F](parsingCost).attempt
```

Comment:
There's been a comment from @JosDenmark about maybe making ParBuilder use `_cost` itself. Given how often we use that in tests, I'd rather leave it as is and handle the single production usage inside the interpreter - as is being done now. I certainly will be against spawning cost instances whenever one needs to parse some rholang. Unless you we hide it under a reasonable util ofc.

Reply:
This will happen the latest when we replace the parsing cost straw man with a proper mechanism, so such a refactor should be executed as a prerequisite PR